src/Shared/Microsoft.AspNetCore.Razor.Test.Common/Language/SyntaxTreeVerifier.cs

```diff
@@ -66,12 +67,87 @@ public static void Verify(RazorSyntaxTree syntaxTree, bool ensureFullFidelity =
+        var firstToken = root.GetFirstToken(includeZeroWidth: true);
+        Assert.Equal(SyntaxKind.EndOfFile, lastToken.Kind);
+        Assert.Null(lastToken.GetNextToken(includeZeroWidth: true));
+        Assert.Null(lastToken.GetNextToken(includeZeroWidth: false));
+        Assert.Null(firstToken.GetPreviousToken(includeZeroWidth: true));
+        Assert.Null(firstToken.GetPreviousToken(includeZeroWidth: false));
+
+        Assert.Same(tokens[0], firstToken);
+        Assert.Same(tokens[^1], lastToken);
+
+        if (tokens.Count == 1)
+        {
+            Assert.Same(lastToken, firstToken);
+            Assert.Null(lastToken.GetPreviousToken(includeZeroWidth: true));
+            Assert.Null(lastToken.GetPreviousToken(includeZeroWidth: false));
+            return;
+        }
+
+        for (var i = 1; i < (tokens.Count - 1); i++)
+        {
+            var previousTokenIndex = i - 1;
+            var previous = tokens[previousTokenIndex];
+            var current = tokens[i];
+            Assert.Same(previous.GetNextToken(includeZeroWidth: true), current);
+            Assert.Same(current.GetPreviousToken(includeZeroWidth: true), previous);
+            validateNonZeroWidth(previous.GetNextToken(includeZeroWidth: false), previousTokenIndex, countUp: true, in tokens);
+            validateNonZeroWidth(previous.GetPreviousToken(includeZeroWidth: false), previousTokenIndex, countUp: false, in tokens);
+        }
+
+        validateNonZeroWidth(lastToken.GetPreviousToken(includeZeroWidth: false), tokens.Count - 1, countUp: false, in tokens);
+
+        void validateNonZeroWidth(SyntaxToken foundNonZeroWidthToken, int originalTokenIndex, bool countUp, in PooledArrayBuilder<SyntaxToken> tokens)
+        {
+            var (targetIndex, increment) = countUp ? (tokens.Count, 1) : (-1, -1);
+            if (foundNonZeroWidthToken == null)
+            {
+                for (var i = originalTokenIndex + increment; i != targetIndex; i += increment)
+                {
+                    Assert.Equal(0, tokens[i].Width);
+                }
+
+                return;
+            }
+
+            Assert.NotEqual(0, foundNonZeroWidthToken.Width);
+
+            for (var i = originalTokenIndex + increment; i != targetIndex; i += increment)
+            {
+                var token = tokens[i];
+
+                if (token.Width == 0)
+                {
+                    continue;
+                }
+
+                Assert.Same(foundNonZeroWidthToken, token);
+                return;
+            }
+
+            Assert.Fail("Did not find the non-zero width token in the list of tokens.");
+        }
     }
 
-    private class Verifier : SyntaxWalker
+    private class Verifier : SyntaxWalker, IDisposable
     {
         private readonly RazorSourceDocument _source;
         private SourceLocation _currentLocation;
+#pragma warning disable CA1805
+        internal PooledArrayBuilder<SyntaxToken> AllTokens = new();
```

Comment:
Since this is for tests anyway, consider an `ImmutableArray<SyntaxToken>.Builder` (or even a plain 'ol `List<SyntaxToken>`) rather than disabling the `PooledArrayBuilder` "non-copyable" warnings. If you need `PooledArrayBuilder` to be a field, please add a comment to advertise that you "know what you are doing". :smile:

Reply:
This isn't a non-copyable warming. It's a warning saying that I shouldn't use `new()` since that's the default, but I prefer the explicitness for this particular type. I'll add a comment.

Reply:
Interesting. That sounds like analyzers aren't running on this particular project.

Reply:
They are, they made sure I didn't accidentally introduce a bug when I typed `readonly` out of habit. There are no copies of the builder in this PR 🙂.

Reply:
That being said, I can move to just List if you prefer. I was mildly concerned about allocations over thousands of tests, but it's probably fine.

Reply:
Huh. I guess there aren't any copies. :smile: I'm OK with leaving it if you think there's a concern.